### PR TITLE
test(collaborate): add collaborate tests

### PR DESCRIPTION
**Pull request by KennethBWSong**
test coverage:
image

### CODE REVIEW

File: packages/fx-core/tests/core/hooks.test.ts [hunk 1]
```diff
@@ -875,4 +891,183 @@ describe("Middleware", () => {
...
+    afterEach(() => {
+      sandbox.restore();
+    });
+
+    it("Previous context and userdata", async () => {
+      envJson = MockPreviousContext();
+      userData = MockPreviousUserData();
+      MockFunctions();
+
+      class MyClass {
```
Comment (SLdragon) [inline packages/fx-core/tests/core/hooks.test.ts:947]: Can we give a meaningful class name and name property? Such as UpgradeHook or something else?

File: packages/fx-core/tests/core/utils.ts [hunk 2]
```diff
@@ -404,3 +405,58 @@ export function MockProjectSettings(appName: string): ProjectSettings {
...
+      local_clientSecret: "{{fx-resource-aad-app-for-teams.local_clientSecret}}",
+      local_objectId: "local_objectId",
+      local_oauth2PermissionScopeId: "local_oauth2PermissionScopeId",
+      local_tenantId: "local_tenantId",
+      local_applicationIdUris: "local_applicationIdUris",
+    },
+  };
+}
+
+export function MockPreviousUserData(): Record<string, string> {
```
Comment (SLdragon) [inline packages/fx-core/tests/core/utils.ts:426]: How about MockPreviousVersionBefore2_3_0UserData, to make it more clear

File: packages/fx-core/tests/core/utils.ts [hunk 3]
```diff
@@ -404,3 +405,58 @@ export function MockProjectSettings(appName: string): ProjectSettings {
...
+  return {
+    "fx-resource-aad-app-for-teams.local_clientSecret": "local_clientSecret",
+  };
+}
+
+export function MockLatestContext(): Json {
```
Comment (SLdragon) [inline packages/fx-core/tests/core/utils.ts:432]: How about MockLatestVersion2_3_0Context, to make it more clear

File: packages/fx-core/tests/core/utils.ts [hunk 4]
```diff
@@ -404,3 +405,58 @@ export function MockProjectSettings(appName: string): ProjectSettings {
...
+      local_objectId: "{{fx-resource-aad-app-for-teams.local_objectId}}",
+      local_oauth2PermissionScopeId:
+        "{{fx-resource-aad-app-for-teams.local_oauth2PermissionScopeId}}",
+      local_tenantId: "{{fx-resource-aad-app-for-teams.local_tenantId}}",
+      local_applicationIdUris: "{{fx-resource-aad-app-for-teams.local_applicationIdUris}}",
+    },
+  };
+}
+
+export function MockLatestUserData(): Record<string, string> {
```
Comment (SLdragon) [inline packages/fx-core/tests/core/utils.ts:450]: same here

File: packages/fx-core/tests/core/hooks.test.ts (hunk 1 above)
Comment (KennethBWSong) [reply packages/fx-core/tests/core/hooks.test.ts]: Fixed. Thanks!

File: packages/fx-core/tests/core/utils.ts (hunk 2 above)
Comment (KennethBWSong) [reply packages/fx-core/tests/core/utils.ts]: Fixed. Thanks!

File: packages/fx-core/tests/core/utils.ts (hunk 3 above)
Comment (KennethBWSong) [reply packages/fx-core/tests/core/utils.ts]: Fixed. Thanks!

File: packages/fx-core/tests/core/utils.ts (hunk 4 above)
Comment (KennethBWSong) [reply packages/fx-core/tests/core/utils.ts]: Fixed. Thanks!

**Review by SLdragon — APPROVED** [review]

**Review by adashen — APPROVED** [review]

**Pull request merged**